components/ide/code/codehelper/main.go

```diff
@@ -0,0 +1,205 @@
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package main
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/sys/unix"
+	"google.golang.org/grpc"
+	yaml "gopkg.in/yaml.v2"
+
+	"github.com/gitpod-io/gitpod/common-go/log"
+	gitpod "github.com/gitpod-io/gitpod/gitpod-protocol"
+	supervisor "github.com/gitpod-io/gitpod/supervisor/api"
+)
+
+var (
+	// ServiceName is the name we use for tracing/logging.
+	ServiceName = "codehelper"
+	// Version of this service - set during build.
+	Version = ""
+)
+
+func main() {
+	log.Init(ServiceName, Version, true, false)
+	startTime := time.Now()
+	log.Info("wait until content available")
+
+	// wait until content ready
+	contentStatus, wsInfo, err := resolveWorkspaceInfo(context.Background())
+	if err != nil || wsInfo == nil || contentStatus == nil || !contentStatus.Available {
+		log.WithError(err).WithField("wsInfo", wsInfo).WithField("cstate", contentStatus).Error("resolve workspace info failed")
+		return
+	}
+	log.WithField("cost", time.Now().Local().Sub(startTime).Milliseconds()).Info("content available")
+
+	code := "/ide/bin/gitpod-code"
+	// start code server with --install-extension flag
+	args := []string{code, "--start-server"}
+
+	if os.Getenv("SUPERVISOR_DEBUG_ENABLE") == "true" {
+		args = append(args, "--inspect", "--log=trace")
+	}
+	wsContextUrl := wsInfo.GetWorkspaceContextUrl()
+	if ctxUrl, err := url.Parse(wsContextUrl); err == nil {
+		if ctxUrl.Host == "github.com" {
+			log.Info("ws context url is from github.com, install builtin extension github.vscode-pull-request-github")
+			args = append(args, "--install-builtin-extension", "github.vscode-pull-request-github")
+		}
+	} else {
+		log.WithError(err).WithField("wsContextUrl", wsContextUrl).Error("parse ws context url failed")
+	}
+
+	uniqMap := map[string]struct{}{}
+	extensions, _ := getExtensions(wsInfo.GetCheckoutLocation())
+	for _, ext := range extensions {
+		if _, ok := uniqMap[ext]; ok {
+			continue
+		}
+		uniqMap[ext] = struct{}{}
+		args = append(args, "--install-extension", ext)
+	}
+
+	args = append(args, os.Args...)
+	log.WithField("code", code).WithField("args", args).
+		WithField("cost", time.Now().Local().Sub(startTime).Milliseconds()).
+		Info("run cmd")
+	args = append([]string{code}, args...)
+
+	// use unix.Exec to replace the calling executable in the process tree
+	if err := unix.Exec(code, args, os.Environ()); err != nil {
+		log.WithError(err).Error("unix exec code failed")
+	}
+}
+
+func resolveWorkspaceInfo(ctx context.Context) (*supervisor.ContentStatusResponse, *supervisor.WorkspaceInfoResponse, error) {
+	resolve := func(ctx context.Context, attempt int) (contentStatus *supervisor.ContentStatusResponse, wsInfo *supervisor.WorkspaceInfoResponse, err error) {
+		supervisorAddr := os.Getenv("SUPERVISOR_ADDR")
```

Comment:
we should be resilient here for both supervisor calls, try like 3 times for everything at least or maybe just forever with a delay